src/services/FeedDataProcessor.ts

```diff
@@ -0,0 +1,203 @@
+import {Logger} from 'winston';
+import {MD5 as hash} from 'object-hash';
+import {inject, injectable} from 'inversify';
+import {LeafValueCoder} from '@umb-network/toolbox';
+
+import MultiFeedProcessor from './FeedProcessor/MultiFeedProcessor';
+import {CalculatorRepository} from '../repositories/CalculatorRepository';
+import {FeedFetcherRepository} from '../repositories/FeedFetcherRepository';
+import Feeds, {FeedCalculator, FeedFetcher, FeedPrice, FeedsInputFetcher, FeedsInputHash} from '../types/Feed';
+import {FeedDatum} from '../types/Datum';
+import {FetcherError} from '../types/Fetcher';
+import {ICalculator} from '../types/Calculator';
+
+@injectable()
+class FeedDataProcessor {
+  @inject('Logger') logger!: Logger;
+  @inject(MultiFeedProcessor) multiFeedProcessor!: MultiFeedProcessor;
+  @inject(CalculatorRepository) calculatorRepository!: CalculatorRepository;
+  @inject(FeedFetcherRepository) feedFetcherRepository!: FeedFetcherRepository;
+
+  async apply(timestamp: number, ...feedsArray: Feeds[]): Promise<{data: FeedDatum[]; prices: FeedPrice[]}> {
+    this.logger.debug('[FeedDataProcessor] #apply');
+
+    const uniqueFeedFetcherMap = this.createUniqueFeedFetcherMap(feedsArray);
+    const {singleInputs, multiInputs} = this.separateInputs(uniqueFeedFetcherMap);
+    const inputIndexByHash = this.createInputByHash(singleInputs, multiInputs);
+    const values = (await this.getProcessedFeeds(singleInputs, multiInputs, timestamp)) as FeedsInputHash;
+    const {data, prices} = this.calculateDataFromFeeds(feedsArray, values, inputIndexByHash, timestamp);
+
+    return {data, prices};
+  }
+
+  private createUniqueFeedFetcherMap(feedsArray: Feeds[]): FeedsInputFetcher {
+    const uniqueFeedFetcherMap: FeedsInputFetcher = {};
+
+    feedsArray.forEach((feeds: Feeds) => {
+      const keys = Object.keys(feeds);
+
+      keys.forEach((leafLabel) =>
+        feeds[leafLabel].inputs.forEach((input) => {
+          uniqueFeedFetcherMap[hash(input.fetcher)] = input.fetcher;
```

Comment:
why we need to create hash if we only using one piece of data? hash will be unique only if fetcher is unique, so we don't have to spend power to compute hash.